module/zfs/dbuf.c

```diff
@@ -1969,31 +2008,7 @@ dbuf_dirty(dmu_buf_impl_t *db, dmu_tx_t *tx)
 	    db->db_state == DB_CACHED || db->db_state == DB_FILL ||
 	    db->db_state == DB_NOFILL);
 
-	mutex_enter(&dn->dn_mtx);
-	/*
-	 * Don't set dirtyctx to SYNC if we're just modifying this as we
-	 * initialize the objset.
-	 */
-	if (dn->dn_dirtyctx == DN_UNDIRTIED) {
-		if (dn->dn_objset->os_dsl_dataset != NULL) {
-			rrw_enter(&dn->dn_objset->os_dsl_dataset->ds_bp_rwlock,
-			    RW_READER, FTAG);
-		}
-		if (!BP_IS_HOLE(dn->dn_objset->os_rootbp)) {
-			dn->dn_dirtyctx = (dmu_tx_is_syncing(tx) ?
-			    DN_DIRTY_SYNC : DN_DIRTY_OPEN);
-			ASSERT(dn->dn_dirtyctx_firstset == NULL);
-			dn->dn_dirtyctx_firstset = kmem_alloc(1, KM_SLEEP);
```

Comment:
I remember thinking, at the time, that this allocation seemed kind of dumb.  However, I suppose the purpose of this allocation was if you had some sort of allocation stack tracking mechanism enabled, you could see how this was reached?  I'm not sure if removing this is an issue for others, especially given one could also use DTrace.

Reply:
That matches my understanding of why the allocation is done here.  Removing this isn't an issue on the Linux side since it's not something we've every made use of.  It also looks to me like `dn_dirtyctx_firstset` and `dn_dirtyctx` are only ever used/checked in ASSERTs.  I don't see any reason this functionality couldn't be moved under `#ifdef ZFS_DEBUG` to make it clear its primary use is debugging  during development.